### ISSUE TITLE
Feat: add a global finalizer callback

### ISSUE THREAD

**Issue by ankostis** [tyarkoni/transitions#186]
After #184 has been merged, this library would have a *global preparation* function but not a corresponding *global finalizer* that would fire once, both for successful/failed transitions, something necessary when implementing protocols like 2-phase commits.  I believe that there is no way to workaround such functionality with code external to this library in a single place (i.e. without proliferating `try...except/finally` in all trigger-calls).

Therefore I suggest to add a `finalize_after_all_transitions` callback in a finally block [inside `Event._trigger()`](/tyarkoni/transitions/blob/c37f5a6c69f8dd8e04e2fe2c5bcbd775c211b479/transitions/core.py#L289-L296).   The code for such a feature, *without error handling* would be something like this:

```python
        for func in self.machine.prepare_conditions_check:
            self.machine._callback(func, event_data)
            logger.debug("Executed machine preparation callback '%s' before conditions." % func)
        ok = False
        try:
            for t in self.transitions[state.name]:
                event_data.transition = t
                if t.execute(event_data):
                    return True
            return False
        finally:
            event_data.kwargs['ok'] = ok  # Provisional name for `ok` flag, or an attr on machine?
            for func in self.machine.finalize_after_all_transitions:
                self.machine._callback(func, event_data)
                logger.debug("Executed machine finalize callback '%s' after all transitions." % func)
```

Of course if *finalize* cb is to receive the error raised, an `except` case is need additionally in the `try` block to store the captured exception.  Like the `ok` flag above, I'm thinking of 2 possibilities where to store it: 

a) on the machine as an extra attribute, or 
b) on the `event-data`, as another keyword item e.g. `error`.  

Case (a) needs clearing this attribute before every trigger and kind of violate separation-of-concerns law; case (b) might clash with names already used by existing clients of the library.

**Comment by aleneum** [comment]
Hi @ankostis,

I added a global `finalize_event` attribute and renamed `prepare_conditions_check` to `prepare_transition`. For now, return values are handled like before: 

If a transition is successful, `Event.trigger` will return `True` and `False` if no transition was applicable (if events are processed instantly and not `queued`).
If a callback raises an exception, this exception will be passed.

The second part of your suggestion is a bit more tricky: Should the result of the transition be passed to `finalize` callbacks and how? Personally, I would prefer to let the user handle exceptions. Also I'd prefer to handle `finalize` callbacks the same was as other callbacks.
As mentioned before, checking the success of a transition can already be done with `Event.trigger`. Silently adding a keyword may lead to errors if callbacks do not expect it.

**Comment by ankostis** [comment]
So with #184, you have implemented a "finalizer" capable of releasing "resources."  

> Personally, I would prefer to let the user handle exceptions. 

I understand that handling exceptions becomes a burden. so you prefer to shift that to the user.
But handling exceptions is a requirement when emulating commit/rollback  behavior.

With #184 as is, a lost of user effort would be needed: capture exceptions inside ALL callbacks, store an appropriate flag somewhere, and then re-raise them.  At least that is my understanding.
Are we sure this is desired?

[brainstorming now] should'nt the true/false trigger flag be an attribute of the event_data?

**Comment by aleneum** [comment]
> capture exceptions inside ALL callbacks

Maybe I am missing something but as far as I can tell the effort stays exactly the same. Currently, a user has to try/except every call to `Event.trigger` if this may raise an exception (`MachineError` or exceptions caused by the callbacks).
This behaviour is not changed because of #184. If you have a variety of callbacks and all of them may raise an exception, you still can do

```python
try:
    model.work()  
except SomeException as e:
    # handle error
```

> shouldn't the true/false trigger flag be an attribute of the event_data?
As mentioned before, adding it to `event_data.kwargs` may lead to errors if callback do not except an automatically generated parameter. Adding `result` to event_data itself could be done without side effects. This requires a Machine with `send_event=True` though.

**Comment by ankostis** [comment]
> [...] the effort stays exactly the same. [...] This behaviour is not changed because of #184.

I believe there is a subtle misunderstanding.
Surrounding `model.work()` won't help you implement commit/rollback behavior WITH `finalize_event` callbacks.

So if (one of) the point of #184 is to support commit/rollback capabilities, then there is no comparison of the effort; previously you just couldn't implement it *centrally*, unless you surrounded every trigger point, as you said.

Is it now more clear what I'm thinking?


> Adding result to event_data itself could be done without side effects. This requires a Machine with send_event=True though.

That could be an interesting opportunity: if you set `send_event=true`, you get exception capturing for free?

**Comment by ankostis** [comment]
(I'll give it another shot)

Before #184, in order to have a resource-cleanup-finalizer, you had to surround every trigger-point in `try/except`.
Now with `finalize_event` you don't have to do that. And that's great.

But you still need this `try/except` for implementing commit/rollback.

**Comment by aleneum** [comment]
Hi @ankostis,

> Surrounding model.work() won't help you implement commit/rollback behavior WITH finalize_event callbacks.

I see. Maybe we can find a way to process errors in callbacks without 'swallowing' caught exceptions in `Event.trigger`. As much as I like to enable your workflow, I also want to keep the current way of doing things for the sake of stability. I do not want to claim that the current way is superior but without further feedback, we cannot distinguish what is more desirable.

> That could be an interesting opportunity: if you set send_event=true, you get exception capturing for free?

Unfortunately, no. Adding the boolean `result` would just tell you if a transition took place. But this could also be achieved with model callbacks as mentioned above.

Perhaps, we can solve this issue on the model level. This way we can keep `core` universal and do not have to introduce another attribute like `raise_exceptions` and/or implement specific finalize callback behaviour.

**Comment by aleneum** [comment]
If you can catch exceptions raised by `Event.trigger` but want to process them in finalize callbacks, we could add `error` to event_data but raise every caught error. This would keep things as they currently are but also add useful information for commit/rollback strategies with finalize. Would that work for you?

**Comment by ankostis** [comment]
You mean a boolean `EventData.error` attribute?
Yes, that's fine.
Actually much better than the OK flag in `kwargs`.

**Comment by ankostis** [comment]
Then it should be called `errored`,  no?

**Comment by aleneum** [comment]
> You mean a boolean EventData.error attribute?

No, I mean the actual error.

Like 

```python
event_data.error = None
try:
...
except Error as e:
    event_data.error = e
    raise e
finally:
    for func in self.machine.finalize_after_all_transitions:
    ...
```

**Comment by ankostis** [comment]
Yes, that would be the a perfect solution.  Thanks.

**Comment by aleneum** [comment]
resolved by #175 I hope
